distributed/diskutils.py

```diff
@@ -115,21 +116,46 @@ class WorkSpace:
     this will be detected and the directories purged.
     """
 
+    base_dir: str
+    _global_lock_path: str
+    _purge_lock_path: str
+
     # Keep track of all locks known to this process, to avoid several
     # WorkSpaces to step on each other's toes
     _known_locks: ClassVar[set[str]] = set()
 
-    def __init__(self, base_dir):
-        self.base_dir = os.path.abspath(base_dir)
-        self._init_workspace()
+    def __init__(self, base_dir: str):
+        self.base_dir = self._init_workspace(base_dir)
         self._global_lock_path = os.path.join(self.base_dir, "global.lock")
         self._purge_lock_path = os.path.join(self.base_dir, "purge.lock")
 
-    def _init_workspace(self):
-        try:
-            os.mkdir(self.base_dir)
-        except FileExistsError:
-            pass
+    def _init_workspace(self, base_dir: str) -> str:
+        """Create base_dir if it doesn't exist.
+        If base_dir already exists but it's not writeable, change the name.
+        """
+        base_dir = os.path.abspath(base_dir)
+        try_dirs = [base_dir]
+        # Note: can't use WINDOWS constant as it upsets mypy
+        if sys.platform != "win32":
+            # - os.getlogin() raises OSError on containerized environments
+            # - os.getuid() does not exist in Windows
+            try_dirs.append(f"{base_dir}-{os.getuid()}")
+
+        for try_dir in try_dirs:
+            try:
+                os.makedirs(try_dir)
+            except FileExistsError:
+                try:
+                    with tempfile.TemporaryFile(dir=try_dir):
+                        pass
+                except PermissionError:
+                    continue
+            return try_dir
+
+        # If we reached this, we're likely in a containerized environment where /tmp
+        # has been shared between containers through a mountpoint, every container
+        # has an external $UID, but the internal one is the same for all.
+        return tempfile.mkdtemp(prefix=base_dir + "-")
```

Comment:
In the low-percentage case that we get here, is it possible that a worker that is killed without running normal cleanup will leave behind directories that we never cleanup when launching new workers? Is that something to worry about?

Reply:
You need to kill -9 the nanny to leave garbage behind, or not have a nanny and kill -9 the worker. In all other cases there will be cleanup.

But yes, this can leave garbage behind if
- you're trigger-happy on SIGKILL, *and*
- you violate segregation by sharing a /tmp mountpoint among containers, *and*
- you do not override the local directory to a local SSD.

Also note that your spill performance will be abysmal if you use a network filesystem for it.

All this together makes me unconcerned about the use case.